integration-tests/tests/delete-orphaned-workspaces.js

```diff
@@ -0,0 +1,82 @@
+const olderThanCount = 12;
+
+/**
+ * Attempts to delete any workspaces which are named with the auto-generated test workspace name prefix,
+ * and are more than a certain configured age. Not a test, a cleanup utility.
+ */
+const deleteOrphanedWorkspaces = withUserToken(async ({ page, testUrl, token }) => {
+  // Sign into Terra so we have the correct credentials.
+  await signIntoTerra(page, { token, testUrl });
+
+  // Delete old orphaned workspaces. These can result from local integration test runs where the
+  // process was prematurely terminated, but a few also appear in alpha and staging indicating that
+  // ci test runs also sometimes leak workspaces.
+  const workspaces = await page.evaluate(async () => await window.Ajax().Workspaces.list());
+  const testWorkspaces = workspaces.filter(({ workspace: { name } }) => _.startsWith(testWorkspaceNamePrefix, name));
+  const testWorkspaceNames = testWorkspaces.map(({ workspace: { name } }) => name).join(', ');
+  console.log(`${testWorkspaces.length} test workspaces (with prefix "${testWorkspaceNamePrefix}") found: ${testWorkspaceNames}`);
+
+  const oldWorkspaces = testWorkspaces.filter(({ workspace: { createdDate } }) => {
+    const getTimeDifference = differenceFnByUnit[timeUnit];
+    const age = getTimeDifference(new Date(createdDate), new Date());
+    return age > olderThanCount;
+  });
+  console.log(`Attempting to delete ${oldWorkspaces.length} test workspaces created more than ${olderThanCount} ${timeUnit} ago.`);
+
+  return Promise.all(
+    _.map(async ({ workspace: { namespace, name, cloudPlatform, state } }) => {
+      try {
+        if (state === 'DeleteFailed') {
+          // Warn before deleting a workspace in a bad state
+          console.log(`Old workspace ${name} is in state 'DeleteFailed'; delete may not succeed.`);
+        }
+        // Unlock the workspace in case it was left in a locked state during a test (locked workspaces can't be deleted).
+        await page.evaluate(async (namespace, name) => await window.Ajax().Workspaces.workspace(namespace, name).unlock(), namespace, name);
+        const isDeleted = await deleteWorkspaceV2({ page, billingProject: namespace, workspaceName: name });
+        return { name, cloudPlatform, isDeleted };
+      } catch (e) {
+        return { name, cloudPlatform, isDeleted: false };
+      }
+    }, oldWorkspaces)
+  ).then((results) => {
+    const deletedNames =
+      results
+        .filter(({ isDeleted }) => isDeleted)
+        .map(({ name, cloudPlatform }) => `${name} (${cloudPlatform})`)
+        .join(', ') || '(none)';
+    const failedNames = results
+      .filter(({ isDeleted }) => !isDeleted)
+      .map(({ name, cloudPlatform }) => `${name} (${cloudPlatform})`)
+      .join(', ');
+    console.info(`Triggered delete on workspaces: ${deletedNames}`);
+    if (failedNames) {
+      console.warn(`Failed to delete workspaces: ${failedNames}`);
```

Comment:
does this occur frequently enough to perhaps warrant a message to a notify channel? If the number is non-zero I'd imagine we need some sort of strategy to ensure we look into it occasionally. Is there any way to have the message/some sort of comment to lead anyone in the future on the right path to be able to manually clean these up?

Reply:
Failures in delete-orphaned-workspaces publish to `dsp-workspaces-test-alerts`. It wasn't set up to fail the test on failure to delete workspaces though - I don't want to make that change in this PR.

Reply:
Realized how to do this in a way that won't error out all the time. Only failing the test if a workspace *newly* transitions to DeleteFailed during the run; existing ones won't trigger failure. So we'll get heads-ups in Slack but won't fail runs forever if a bad penny turns up.

Reply:
Yes, please don't put a Slack notification for old borked workspaces that keep failing to delete. The purpose of this cleanup was to prevent the number of orphaned workspaces from continuously increasing (which is the situation we were in when I wrote it), not to test workspace deletion itself.